### PR TITLE
Readd cooldowns on the world topic

### PR DESCRIPTION
**Pull request by optimumtact**
Thanks to whomever removed these, you're a real star

### CODE REVIEW

**Comment by deathride58** [comment]
putting the cooldowns/spam checks prior to `TGS_TOPIC` means that all of the relevant spam checks will essentially be bypassed, since most of the things that respond to topic calls are handled via `TgsTopic()`

**Comment by optimumtact** [comment]
I actually mean to move that and forgot

**Comment by optimumtact** [comment]
@Cyberboss @MrStonedOne

**Comment by MrStonedOne** [comment]
This would trigger for the webserver, it requests topic at a rate of once per second.

File: code/game/world.dm [hunk 1]
```diff
@@ -164,10 +166,30 @@ GLOBAL_VAR(restart_counter)
...
 /world/Topic(T, addr, master, key)
-	TGS_TOPIC	//redirect to server tools if necessary
-
+	var/static/list/bannedsourceaddrs = list()
+	var/static/list/lasttimeaddr = list()
 	var/static/list/topic_handlers = TopicHandlers()
 
+	//LEAVE THIS COOLDOWN HANDLING IN PLACE, OR SO HELP ME I WILL MAKE YOU SUFFER
+	if (addr in bannedsourceaddrs)
+		return
```
Comment (MrStonedOne) [inline code/game/world.dm]: Use an associated list reference instead of `in`.

`in` is O(n), `blah[key]` is O(log n)

Comment (optimumtact) [reply code/game/world.dm]: for some reason I thought that gives a runtime about the key if it doesn't exist

**Comment by AndriiYukhymchak** [comment]
> This would trigger for the webserver, it requests topic at a rate of once per second.

I think if you add TGS to some sort of IP whitelist...

**Comment by Cyberboss** [comment]
All that would be needed is localhost

File: code/game/world.dm [hunk 2]
```diff
@@ -164,10 +166,30 @@ GLOBAL_VAR(restart_counter)
...
+		bannedsourceaddrs[addr] = TOPIC_BANNED
+		return
+	if (addr in lasttimeaddr)
+		var/lasttime = lasttimeaddr[addr]
+		if(world.time < (lasttime + 2 SECONDS))
+			log_admin_private("[addr] banned from topic calls for a round for too frequent messages")
+			bannedsourceaddrs[addr] = TOPIC_BANNED
+			return
+
+	lasttimeaddr[addr] = world.time
```
Comment (AndriiYukhymchak) [inline code/game/world.dm]: instead use 
`
lasttimeaddr[addr] = world.time + 2 SECONDS
`
and check that 
`
if(world.time < lasstime)
`

**Review by NethIafin — CHANGES_REQUESTED** [review]

**Comment by MrStonedOne** [comment]
>I think if you add TGS to some sort of IP whitelist...

The webserver/game banners do not use TGS.

They do not come from localhost. they come from a lan ip for all servers except the eu servers, those see the internet ip of the webserver.

**Comment by MrStonedOne** [comment]
This needs to be configurable, and should either be the normal standard format of count in time, or a better count in static timeframe  system.

Look at the topic limiter for an example of a robust anti-spam system.

The ability to add ips on a whitelist would also help.

**Comment by AndriiYukhymchak** [comment]
this doesn't appear to help. Tested close enough implementation on CM, stayed dead.

**Comment by Crossedfall** [comment]
Would still be good to have regardless. As for the whitelisting, could just open up a new config entry for that? That way you don't have to hardcode IPs into the handling.

**Pull request closed without merging**